### ISSUE TITLE
atom guesser incorrectly warns about mass==0 atom types

### ISSUE THREAD

**Issue by orbeckst** [MDAnalysis/mdanalysis#1689]
### Expected behaviour
Assigning a dummy atom (`M`) with mass 0 should not raise a warning.

### Actual behaviour
A warning is raised even though the atom type was correctly assigned:
```
/home/richard/miniconda2/envs/mdadev/lib/python2.7/site-packages/MDAnalysis/topology/guessers.py:56: UserWarning: Failed to guess the mass for the following atom types: M
  "".format(', '.join(misses)))
```

### Code to reproduce the behaviour

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import GRO

u = mda.Universe(GRO)
```
gives the warning

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
0.17.0-dev

**Comment by apoorvaeternity** [comment]
@orbeckst I want to work on this issue. Please give me a glimpse on how to proceed.

**Comment by orbeckst** [comment]
Hi @apoorvaeternity , welcome to MDAnalysis!

## preparation
Have you *used* MDAnalysis before? If not I suggest you spend a little time to familiarize yourself with the library and the underlying ideas – otherwise it's kind of hard to make sense of some of the jargon that is used (such as "topology" and "atom guessers" in this issue). As an introduction I recommend
- read  [our SciPy2016 paper](http://conference.scipy.org/proceedings/scipy2016/oliver_beckstein.html) (and you can also watch the [video of my SciPy presentation on MDAnalysis](https://youtu.be/zVQGFysYDew))
- work through the [MDAnalysis Tutorial](https://www.mdanalysis.org/MDAnalysisTutorial/)
- have a look at our [Developer Guide](https://github.com/MDAnalysis/mdanalysis/wiki/Guide-for-Developers)

## improving the atom guesser
We can then look more closely into `topology/guessers.py`. At the moment, the code assumes that a mass of 0 is a sign that the atom type was not assigned. https://github.com/MDAnalysis/mdanalysis/blob/06fae8b6c33f6c8763779b772a6cf2f4904bf5ac/package/MDAnalysis/topology/guessers.py#L52

Instead we need a check if there is actually a type assigned. However, this needs to be done in a way as to not slow down the loading process (which is already slow due to atom type guessing). Furthermore, the way things are currently structured, there is little information communicated into the function.

We would  proceed by you 
1. [writing a test](https://github.com/MDAnalysis/mdanalysis/wiki/UnitTests) for this issue
2. proposing an initial solution (discuss briefly here on the issue tracker, then submit a PR)

We then continue the discussion on the PR and refine the solution. It then gets formally reviewed, and if approved, your PR will get merged (see [Contributing code](https://github.com/MDAnalysis/mdanalysis/wiki/ContributingCode)).

**Comment by palnabarun** [comment]
@orbeckst Is anyone actively working on this? I looked into the code and have this proposed solution.

Currently, the `topology.guesser.get_atom_mass` return `0.00` if the element name is not found in the dictionary. After inspecting `topology.guesser.guess_atom_element`, I can see that it returns nothing if it is unable to guess the atom element name. What I propose is to raise a warning then and there itself. This will help in catching undefined atom types. What are your thoughts on this?

**Comment by orbeckst** [comment]
No-one has submitted a PR so I consider this an open issue.

Please read the https://github.com/MDAnalysis/mdanalysis/issues/1689#issuecomment-359497868 . Ultimately, we want the warning only to occur if truly no mass could be guessed. If you can achieve this by changing were the warning is raised then that might work. I suggest you submit a PR because hat's the most productive way to discuss code. And it might show other side effects that are not obvious from your verbal description.

**Comment by palnabarun** [comment]
Hi @orbeckst, Can you please give me an example where the warning should ideally occur? In the case mentioned in the issue, the warning shouldn't occur, right?

**Comment by orbeckst** [comment]
We still want to know when we cannot guess a mass correctly, so ideally for that the same place.

> In the case mentioned in the issue, the warning shouldn't occur, right?

Yes, if we can actually get the mass from elsewhere instead of guessing it then no warning should be raised.

**Comment by palnabarun** [comment]
I think that when the label associated with the atom type is not there in our atom mass table, there should be a warning that atom mass cannot be guessed definitively.

Also, as of now, the atom guesser only gets a list with the atom labels like 'H', 'He', 'M' etc. There is no other information associated with the input.

**Comment by orbeckst** [comment]
Yes, that is the challenge.

**Comment by palnabarun** [comment]
I looked into the GRO format and that does not seem to enforce any information for atom masses. It just has labels for the constituent atoms. For GRO, we can not definitively guess unknown atom type masses. As for other formats, I did not deep that much.

For now, can we just raise a warning when the element is not available in our data tables? This is not a perfect solution but is faster than the one right now.

**Comment by orbeckst** [comment]
> I looked into the GRO format and that does not seem to enforce any information for atom masses. It just has labels for the constituent atoms.  For GRO, we can not definitively guess unknown atom type masses. 

Yes, that's the reason why we are *guessing*. 

I *thought* that MDA would recognize "MW" as a dummy. However, I am not so sure anymore. And based on the output from your PR #1773 it also seems to be the case that "M" or "MW" was not recognized in the first place. However, from looking at the code it is clear that as soon as we're guessing (as for GRO but not eg TPR, which contains masses), we run into the issue described here.

Have you looked into adding a line
```
MW      DUMMY
```
to `topology.tables.TABLE_ATOMELEMENTS` https://github.com/MDAnalysis/mdanalysis/blob/2703b38b5fa7f90857d1dd5a3d2a6020861913b2/package/MDAnalysis/topology/tables.py#L79  and then adding a line
```
DUMMY      0.0
```
to  table `topology.TABLE_MASSES` https://github.com/MDAnalysis/mdanalysis/blob/2703b38b5fa7f90857d1dd5a3d2a6020861913b2/package/MDAnalysis/topology/tables.py#L171 

This should allow us to (1) recognize dummy atoms and assign a mass properly and (2) make it easy to add other dummy atom names (there is also "DUM" and many more in addition to "MW").

Once we have this, we can use an approach such as your validation idea in https://github.com/MDAnalysis/mdanalysis/pull/1773#issuecomment-363774446 . Unless we set unknown masses to `None` (bad for arrays), `0j` (bad for arrays), `-100` (ugly but doable, if we fix it later) we will probably have to do something like your explicit validation (which is pretty clear but I don't know what the performance impact is).

**Comment by palnabarun** [comment]
One thing to note. It was bothering me since the first time I saw the code. In the `guess_masses` function,
```
def guess_masses(atom_types):
    """Guess the mass of many atoms based upon their type

    Parameters
    ----------
    atom_types
      Type of each atom

    Returns
    -------
    atom_masses : np.ndarray dtype float64
    """
    masses = np.array([get_atom_mass(atom_t) for atom_t in atom_types], dtype=np.float64)
    if np.any(masses == 0.0):
        # figure out where the misses were and report
        misses = np.unique(np.asarray(atom_types)[np.where(masses == 0.0)])
        warnings.warn("Failed to guess the mass for the following atom types: {}"
                      "".format(', '.join(misses)))
    return masses
```
the `get_atom_mass` function is called. I think that instead of that `guess_atom_mass` function should have been called which actually guesses the element from the atom name. I put up now as in the discussion above, it is clear that `guess_atom_mass` should have been called instead of `get_atom_mass`.

**Comment by ViraRa** [comment]
Does anyone know how to get rid of the warning? 

UserWarning: Failed to guess the mass for the following atom types: M "".format(', '.join(misses)))

I am using version 2.0.0.

**Comment by IAlibay** [comment]
> Does anyone know how to get rid of the warning? 
> 
> UserWarning: Failed to guess the mass for the following atom types: M "".format(', '.join(misses)))
> 
> I am using version 2.0.0. 

@ViraRa Could I ask you to ask this question either in the mailing list (if it's a usage question) or a new issue (if it's a bug / feature request)? It's a lot easier for folks to find your question & answer if it's not on an already closed issue.

**Comment by ViraRa** [comment]
> > Does anyone know how to get rid of the warning?
> > UserWarning: Failed to guess the mass for the following atom types: M "".format(', '.join(misses)))
> > I am using version 2.0.0.
> 
> @ViraRa Could I ask you to ask this question either in the mailing list (if it's a usage question) or a new issue (if it's a bug / feature request)? It's a lot easier for folks to find your question & answer if it's not on an already closed issue.

Sure. I will put it in the mailing list.

**Comment by HsuShihHsueh** [comment]
Has anyone solved this problem?
